### PR TITLE
Add support for disk encryption key in GCPMachine

### PR DESCRIPTION
**Pull request by bfournie**
**What type of PR is this?**

**What this PR does / why we need it**:

Add support for the disk encryption key for the boot disk and and additional disks in the GCP machine.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [x] adds unit tests

**Release note**:

```release-note
Add support for disk encryption key in GCPMachine
```

### CODE REVIEW

File: api/v1beta1/gcpmachine_types.go
Comment (JoelSpeed) [inline api/v1beta1/gcpmachine_types.go:175]: Which of these fields work together, and which do not? Typically in Kube when there's a choice, either supply a reference to a key, or provide the raw key, we use a discriminated union and make the user choose.

Comment (bfournie) [reply api/v1beta1/gcpmachine_types.go:175]: > Which of these fields work together, and which do not? Typically in Kube when there's a choice, either supply a reference to a key, or provide the raw key, we use a discriminated union and make the user choose.

The way I understand it - the encryption can be defined by only one of KmsKeyName, RawKey, or RsaEncryptedKey. KmsKeyServiceAccount is applicable to any of the encryption choices. 
Is there an example of a way to represent this?

Comment (bfournie) [reply api/v1beta1/gcpmachine_types.go:175]: Thank you, will update.

Comment (bfournie) [reply api/v1beta1/gcpmachine_types.go]: Added a KeyType and separate structs to distinguish the settings for CustomerManaged or CustomerSupplied keys. I didn't further split out the RawKey from the RsaEncryptedKey as its documented that either can be provided.

File: api/v1beta1/gcpmachine_types.go
Comment (bfournie) [reply api/v1beta1/gcpmachine_types.go:172]: > Normally this would also be a discriminated union
I added `+kubebuilder:validation:MinProperties=1` for this per comment below.

File: api/v1beta1/gcpmachine_types.go
Comment (bfournie) [reply api/v1beta1/gcpmachine_types.go]: Done.

Comment (bfournie) [reply api/v1beta1/gcpmachine_types.go]: Done.

File: api/v1beta1/gcpmachine_types.go
Comment (bfournie) [reply api/v1beta1/gcpmachine_types.go]: Fixed.

File: api/v1beta1/gcpmachine_types.go
Comment (bfournie) [reply api/v1beta1/gcpmachine_types.go:193]: > You can either use CEL or a webhook, but you should validate that ManagedKey and SuppliedKey are supplied if and only if KeyType matches

Added a webhook.

File: api/v1beta1/gcpmachine_types.go
Comment (JoelSpeed) [inline api/v1beta1/gcpmachine_types.go:167]: Does GCP specify any constraints for the name of the KMS key? Upper case? Lower case? Certain special characters allowed? MInimum or maximum length? All of this could be validated at admission time to prevent errors down the line

File: api/v1beta1/gcpmachine_types.go
Comment (JoelSpeed) [inline api/v1beta1/gcpmachine_types.go:177]: Base64 strings in kube are typically `[]byte` not `*string`. Eg secrets data, or CAData fields in webhooks. IIRC the serializers decode and encode the base64 for you if you're doing this, which should make handling this field easier

File: api/v1beta1/gcpmachine_types.go
Comment (JoelSpeed) [inline api/v1beta1/gcpmachine_types.go:189]: Should be cased as `RSA`
```suggestion
	RSAEncryptedKey *string `json:"rsaEncryptedKey,omitempty"`
```

Comment (JoelSpeed) [inline api/v1beta1/gcpmachine_types.go:189]: As above, should be `[]byte` not `*string`

File: api/v1beta1/gcpmachine_types.go
Comment (JoelSpeed) [inline api/v1beta1/gcpmachine_types.go:197]: Is this allowing the empty string as a value in the enum?

File: api/v1beta1/gcpmachine_types.go
Comment (JoelSpeed) [inline api/v1beta1/gcpmachine_types.go:200]: Could apply validation for this, valid characters/format, min and max length etc

File: api/v1beta1/gcpmachine_types.go
Comment (JoelSpeed) [inline api/v1beta1/gcpmachine_types.go:202]: Optional strings should be pointers according to kube conventions

File: api/v1beta1/gcpmachine_types.go
Comment (JoelSpeed) [inline api/v1beta1/gcpmachine_types.go:208]: Should state (and validate) that these are only permitted when `KeyType` is set to the appropriate value

File: api/v1beta1/gcpmachine_types.go
Comment (JoelSpeed) [inline api/v1beta1/gcpmachine_types.go:320]: Are you ever likely to expand a difference between this use of `CustomerEncryptionKey` and the other one? Typically we recommend not re-using structs as you can accidentally expand an API you didn't intend to, though in this case it may be safe

File: api/v1beta1/gcpmachine_webhook.go
Comment (JoelSpeed) [inline api/v1beta1/gcpmachine_webhook.go:134]: Want to handle KeyType being something other than the other two values? Maybe a switch with a default case of unknown value would be better?

File: api/v1beta1/gcpmachine_types.go
Comment (bfournie) [reply api/v1beta1/gcpmachine_types.go]: Good point, changed to []byte.

File: api/v1beta1/gcpmachine_types.go
Comment (bfournie) [reply api/v1beta1/gcpmachine_types.go]: Fixed.

File: api/v1beta1/gcpmachine_types.go
Comment (bfournie) [reply api/v1beta1/gcpmachine_types.go:197]: Good catch. Removed the trailing ';'

File: api/v1beta1/gcpmachine_types.go
Comment (bfournie) [reply api/v1beta1/gcpmachine_types.go]: Good point, fixed.

File: api/v1beta1/gcpmachine_types.go
Comment (bfournie) [reply api/v1beta1/gcpmachine_types.go]: Added comments for that. The validation should be done.

File: api/v1beta1/gcpmachine_webhook.go
Comment (bfournie) [reply api/v1beta1/gcpmachine_webhook.go:138]: Good idea. Changed.

File: api/v1beta1/gcpmachine_types.go
Comment (bfournie) [reply api/v1beta1/gcpmachine_types.go:320]: In this case, the struct is used for disks only on (either AdditionalDisks or RootDisk) it seems we would want to keep the same encryption capabilities. The AttachedDiskSpec also has DiskType and size, the same as what's defined for RootDeviceType and RootDeviceSize.

File: api/v1beta1/gcpmachine_types.go
Comment (bfournie) [reply api/v1beta1/gcpmachine_types.go]: Changed to []byte.

File: api/v1beta1/gcpmachine_types.go
Comment (bfournie) [reply api/v1beta1/gcpmachine_types.go:167]: So unless I'm missing something I don't see constraints defined in the API https://pkg.go.dev/google.golang.org/api/compute/v1#CustomerEncryptionKey or in validations. I'm leery to add additional limit checks here unless its well defined.

File: api/v1beta1/gcpmachine_types.go
Comment (bfournie) [reply api/v1beta1/gcpmachine_types.go:200]: Yes, still not sure what this should be limited to though.

File: api/v1beta1/gcpmachine_types.go
Comment (JoelSpeed) [reply api/v1beta1/gcpmachine_types.go:167]: When trying to create a key, it gives me
```
Key names can contain letters, numbers, underscores (_), and hyphens (-). Keys can't be renamed or deleted.
```
I believe that is also true of project IDs and regions so we probably can limit to that selection plus the slashes required

Comment (richardcase) [reply api/v1beta1/gcpmachine_types.go:167]: Thanks for testing that @JoelSpeed.

File: api/v1beta1/gcpmachine_types.go
Comment (richardcase) [inline api/v1beta1/gcpmachine_types.go:171]: What happens if the user supplies both the **RawKey** and **RSAEncryptionKey**. Should we add validation to ensure both are not supplied?

File: api/v1beta1/gcpmachine_types.go
Comment (bfournie) [reply api/v1beta1/gcpmachine_types.go:167]: Thanks, I'll add a check in the webhook.Validator.

File: api/v1beta1/gcpmachine_types.go
Comment (bfournie) [reply api/v1beta1/gcpmachine_types.go:171]: Yes, I'll add a check in webhook.Validator.

File: api/v1beta1/gcpmachine_types.go
Comment (JoelSpeed) [reply api/v1beta1/gcpmachine_types.go:167]: Any webhook validations also should be explained in the godoc

Personal quarrel but, webhook validations are hard to review from an API review perspective. We have kubebuilder validations and CEL validations that can handle these validations and live alongside the field, it's much easier to see the validation and understand the behaviour of the field when it's written right next to it.

I know the CAPI projects lean a lot on webhooks and they seem preferred for now, but, I want to raise that point at a future community call

Comment (bfournie) [reply api/v1beta1/gcpmachine_types.go:167]: OK, so would be preferred way for something like this to use:
// +kubebuilder:validation:Pattern
and
// +kubebuilder:validation:MaxLength
?

Comment (JoelSpeed) [reply api/v1beta1/gcpmachine_types.go:167]: I think that's clearer personally yes.

You can also do CEL stuff but I know CAPI aren't doing that much yet, for this one a pattern probably makes sense as it's simple to understand

```
// +kubebuilder:validation:pattern="projects/[-_[a-z0-9]+/locations/region/keyRings/[-_[a-z0-9]+/cryptoKeys/[-_[a-z0-9]+"
```

For the other field you could do:

```
// +kubebuilder:validation:XValidation:rule="self.contains('@')",message="kmsKeyServiceAccount should be an email address"
// +kubebuilder:validation:XValidation:rule="self.split('@')[0].matches('[a-z][-a-zA-Z0-9]{5,29}')",message="kmsServiceAccount email prefix must start with a lower case letter, followed by one or more lower case alphanumerical characters that can be separated by hyphens and must be between 6 and 30 characters in length.
// +kubebuilder:validation:XValidation:rule="self.endsWith('.iam.gserviceaccount.com')",message="kmsKeyServiceAccount must be a valid gcloud service account. Must end with '.iam.gserviceaccount.com'."
```

Will want a maintainers ack on using these though, I'm not a maintainer here

Comment (damdo) [reply api/v1beta1/gcpmachine_types.go:167]: ^ cc. @richardcase @cpanato

Comment (bfournie) [reply api/v1beta1/gcpmachine_types.go:168]: Thanks for the guidance. With some minor changes that first regexp seems to work well, e.g. https://regex101.com/r/7Zjc3O/1

For the kmsKeyServiceAccount, I had to reduce the number of Validation rules because of test failures
`Forbidden: estimated rule cost exceeds budget by factor of more than 100x (try simplifying the rule, or adding maxItems, maxProperties, and maxLength) `. 
I derived the maxlength from the google docs, e.g. https://cloud.google.com/iam/quotas

File: api/v1beta1/gcpmachine_types.go
Comment (bfournie) [reply api/v1beta1/gcpmachine_types.go]: Used a limit of 85 based on limits for service and project id, along with email suffix.

File: api/v1beta1/gcpmachine_types.go
Comment (richardcase) [reply api/v1beta1/gcpmachine_types.go:168]: @JoelSpeed - you make a good point about kubebuilder validations/CEL vs webhook validation. I think part of the current reliance on webhook validation in CAPI is historical because CEL wasn't available for more complex validation logic.

File: api/v1beta1/gcpmachine_types.go
Comment (JoelSpeed) [reply api/v1beta1/gcpmachine_types.go:174]: Should we also add `MaxProperties`, which would give a clear indication that you can have 1, and only 1 of the two options?

File: api/v1beta1/gcpmachine_types.go
Comment (JoelSpeed) [inline api/v1beta1/gcpmachine_types.go:208]: Non validation on the name or project id currently, could be done with CEL or with webhook for now

Comment (bfournie) [reply api/v1beta1/gcpmachine_types.go:209]: Changed to use regexp instead of XValidation as multiple XValidations caused problems. This keeps all of the validations for this string in one place.

File: api/v1beta1/gcpmachine_types.go
Comment (bfournie) [reply api/v1beta1/gcpmachine_types.go:175]: Added `MaxProperties`

File: api/v1beta1/gcpmachine_types.go
Comment (bfournie) [reply api/v1beta1/gcpmachine_types.go:169]: Sounds good. Hopefully we can get this merged and revisit the validations after the discussion.

File: api/v1beta1/gcpmachine_types.go
Comment (JoelSpeed) [reply api/v1beta1/gcpmachine_types.go:209]: Pattern works, curious what problems the multiple xvalidations added though if you have the output to hand

Comment (bfournie) [reply api/v1beta1/gcpmachine_types.go:209]: Using the 3 XValidations had caused this error https://github.com/kubernetes-sigs/cluster-api-provider-gcp/pull/1137#discussion_r1491655487

Comment (JoelSpeed) [reply api/v1beta1/gcpmachine_types.go:209]: I suspect it would work if you include the 3 validations plus the maxlength on the field, from my read of that thread you haven't tried that combination yet

Comment (bfournie) [reply api/v1beta1/gcpmachine_types.go:209]: Even with the `MaxLength` I get an error when using multiple XValidation rules, for example with just 2:
        // +kubebuilder:validation:MaxLength=85
       // +kubebuilder:validation:XValidation:rule="self.contains('@')",message="kmsKeyServiceAccount should be an email address"
       // +kubebuilder:validation:XValidation:rule="self.split('@')[0].matches('[a-z][-a-zA-Z0-9]{5,29}')",message="kmsServiceAccount email prefix must start with a lower case letter, followed by one or more lower case alphanumerical characters that can be separated by hyphens and must be between 6 and 30 characters in length."

I get:
```
  [FAILED] Unexpected error:
      <*fmt.wrapError | 0xc00061e180>: 
      unable to install CRDs onto control plane: unable to create CRD instances: unable to create CRD "gcpmachinetemplates.infrastructure.cluster.x-k8s.io": CustomResourceDefinition.apiextensions.k8s.io "gcpmachinetemplates.infrastructure.cluster.x-k8s.io" is invalid: [spec.validation.openAPIV3Schema.properties[spec].properties[template].properties[spec].properties[additionalDisks].items.properties[encryptionKey].properties[kmsKeyServiceAccount].x-kubernetes-validations[0].rule: Forbidden: estimated rule cost exceeds budget by factor of 3.7x (try simplifying the rule, or adding maxItems, maxProperties, and maxLength where arrays, maps, and strings are declared), spec.validation.openAPIV3Schema.properties[spec].properties[template].properties[spec].properties[additionalDisks].items.properties[encryptionKey].properties[kmsKeyServiceAccount].x-kubernetes-validations[1].rule: Forbidden: estimated rule cost exceeds budget by factor of 29.4x (try simplifying the rule, or adding maxItems, maxProperties, and maxLength where arrays, maps, and strings are declared), spec.validation.openAPIV3Schema.properties[spec].properties[template].properties[spec].properties[additionalDisks].items.properties[encryptionKey].properties[kmsKeyServiceAccount].x-kubernetes-validations[1].rule: Forbidden: contributed to estimated rule cost total exceeding cost limit for entire OpenAPIv3 schema, spec.validation.openAPIV3Schema.properties[spec].properties[template].properties[spec].properties[additionalDisks].items.properties[encryptionKey].properties[kmsKeyServiceAccount].x-kubernetes-validations[0].rule: Forbidden: contributed to estimated rule cost total exceeding cost limit for entire OpenAPIv3 schema, spec.validation.openAPIV3Schema: Forbidden: x-kubernetes-validations estimated rule cost total for entire OpenAPIv3 schema exceeds budget by factor of 3.3x (try simplifying the rule, or adding maxItems, maxProperties, and maxLength where arrays, maps, and strings are declared)]

```

Comment (JoelSpeed) [reply api/v1beta1/gcpmachine_types.go:209]: Cool, thanks for trying, lets continue with the validations we have here for now then

Comment (bfournie) [reply api/v1beta1/gcpmachine_types.go:209]: Thanks

**Review by JoelSpeed — COMMENTED** [review]
/lgtm

**Review by cpanato — APPROVED** [review]
thanks

/approve

/hold for @richardcase as he did some reviews as well

**Review by damdo — COMMENTED** [review]
/lgtm

Let's see what @richardcase thinks

**Comment by richardcase** [comment]
Looks good to me, thanks @bfournie . Good discussion on CEL as part of this :+1: 

/approve

**Comment by damdo** [comment]
I asked @bfournie to rebase this on top of `main` once #1157 merges.
I suspect something in the tooling is breaking (could be the Go version as I've seen this break with Netlify already recently).
So a rebase could help here.

**Review by damdo — COMMENTED** [review]
/lgtm

**Pull request merged**
